src/vma/dev/qp_mgr_eth_mlx5.cpp

```diff
@@ -74,4 +155,154 @@ cq_mgr* qp_mgr_eth_mlx5::init_rx_cq_mgr(struct ibv_comp_channel* p_rx_comp_event
+
+	/*
+	 * Make sure that descriptors are written before
+	 * updating doorbell record and ringing the doorbell
+	 */
+	wmb();
+	*m_sq_db = htonl(m_sq_wqe_cntr);
+
+	/* This wc_wmb ensures ordering between DB record and BF copy */
+	wc_wmb();
+
+	/*
+	 * Avoid using memcpy() to copy to BlueFlame page, since memcpy()
+	 * implementations may use move-string-buffer assembler instructions,
+	 * which do not guarantee order of copying.
+	 */
+	mlx5_bf_copy((volatile uintptr_t *)((uintptr_t)m_sq_bf_reg + m_sq_bf_offset),
+		(volatile uintptr_t *)m_sq_wqe_hot);
+
+	m_sq_bf_offset ^= m_sq_bf_buf_size;
+
+	m_sq_wqe_idx_to_wrid[m_sq_wqe_hot_index] = (uintptr_t)p_send_wqe->wr_id;
+
+	/*Set the next WQE and index*/
+	m_sq_wqe_hot = &(*m_sq_wqes)[m_sq_wqe_cntr & (m_tx_num_wr - 1)];
+	/* Write only data[0] which is the single element which changes.
+	 * Other fields are already initialised in mlx5_init_sq. */
+	m_sq_wqe_hot->ctrl.data[0] = htonl((m_sq_wqe_cntr << 8) | MLX5_OPCODE_SEND);
+	m_sq_wqe_hot_index = m_sq_wqe_cntr & (m_tx_num_wr - 1);
+	return 0;
+}
+
+// Handle releasing of Tx buffers
+// Single post send with SIGNAL of a dummy packet
+
+// NOTE: Since the QP is in ERROR state no packets will be sent on the wire!
+// So we can post_send anything we want :)
+void qp_mgr_eth_mlx5::trigger_completion_for_all_sent_packets()
+{
+	qp_logfunc("unsignaled count=%d, last=%p", m_n_unsignaled_count, m_p_last_tx_mem_buf_desc);
+
+	if (m_p_last_tx_mem_buf_desc) { // Meaning that there is at least one post_send in the QP mem_buf_desc that wasn't signaled for completion
+		qp_logdbg("Need to send closing tx wr...");
+		// Allocate new send buffer
+		mem_buf_desc_t* p_mem_buf_desc = m_p_ring->mem_buf_tx_get(0, true);
+		m_p_ring->m_missing_buf_ref_count--; // Align Tx buffer accounting since we will be bypassing the normal send calls
+		if (!p_mem_buf_desc) {
+			qp_logerr("no buffer in pool");
+			return;
+		}
+		p_mem_buf_desc->p_next_desc = m_p_last_tx_mem_buf_desc;
+
+		// Prepare dummy packet: zeroed payload ('0000').
+		// For ETH it replaces the MAC header!! (Nothing is going on the wire, QP in error state)
+		/* need to send at least eth+ip, since libmlx5 will drop just eth header */
+		ethhdr* p_buffer_ethhdr = (ethhdr *)p_mem_buf_desc->p_buffer;
+		memset(p_buffer_ethhdr, 0, sizeof(*p_buffer_ethhdr));
+		p_buffer_ethhdr->h_proto = htons(ETH_P_IP);
+		iphdr* p_buffer_iphdr = (iphdr *)(p_mem_buf_desc->p_buffer + sizeof(*p_buffer_ethhdr));
+		memset(p_buffer_iphdr, 0, sizeof(*p_buffer_iphdr));
+
+		ibv_sge sge[1];
+		sge[0].length = sizeof(ethhdr) + sizeof(iphdr);
+		sge[0].addr = (uintptr_t)(p_mem_buf_desc->p_buffer);
+		sge[0].lkey = m_p_ring->m_tx_lkey;
+
+		// Prepare send wr for (does not care if it is UD/IB or RAW/ETH)
+		// UD requires AH+qkey, RAW requires minimal payload instead of MAC header.
+		vma_ibv_send_wr send_wr;
+
+		memset(&send_wr, 0, sizeof(send_wr));
+		send_wr.wr_id = (uintptr_t)p_mem_buf_desc;
+		send_wr.wr.ud.ah = NULL;
```

Comment:
The original `trigger_completion_for_all_sent_packets()` (in `qp_mgr.cpp`) has:

```
send_wr.wr.ud.remote_qpn = FICTIVE_REMOTE_QPN;
send_wr.wr.ud.remote_qkey = FICTIVE_REMOTE_QKEY;
```

Why is this excluded from here?